src/main/java/com/zaxxer/nuprocess/linux/ProcessEpoll.java

```diff
@@ -164,6 +182,19 @@ public void queueWrite(LinuxProcess process)
       }
    }
 
+   @Override
+   public void run() {
+      super.run();
+
+      if (process != null) {
+         // For synchronous execution, perform a final _blocking_ deadpool check. If the deadpool
+         // is empty, this will return immediately. If the process has already exited, waitpid will
+         // return immediately. Otherwise, this will block until the the process terminates. This
+         // is necessary to ensure the handler's onExit is called before LinuxProcess.run returns.
+         checkDeadPool(0);
```

Comment:
I don't love this solution, but I'm torn between different options without any that I think are great. I'm keen to get your feedback.

The problem: With async processing, after a given process terminates, the pump thread is still running for some number of linger iterations (or potentially even longer if there are other active processes), and it's periodically probing the `deadPool` for processes that have finished up. That means, eventually, `waitpid` is going to detect that the process ended and `onExit` will be called. For _synchronous_ processes, after this `run()` call returns, the `deadPool` will never be checked again. That means if the process hasn't _fully exited_ by this point, `onExit` will never happen (and the process will become a zombie, since nothing will ever wait for it).

Options I considered:
* Instead of using a -1 for `lingerIterations`, use 0. That effectively requires at least one iteration (since the check is `>` rather than `>=`). This works, but effectively adds a _500ms delay_ to _every synchronous process_ (250ms for an extra loop to check the `deadPool` a second time; this loop will reap the process, but won't be considered "idle", and another 250ms for the loop _after_ that one, to actually show as "idle")
  * That's a huge penalty, especially for something like Bitbucket Server that forks dozens or hundreds of concurrent processes and can fork out several very short-lived processes in rapid succession to complete a single task
* Add `protected` method to `BaseEventProcessor` and update its `run` loop to call that instead of, or in addition to, checking `pidToProcessMap.isEmpty()`, which this class could override to also check `deadPool.isEmpty()` (like `ProcessCompletions` does on Windows)
  * This could drop the penalty as low as 250ms, since I can pair it with `-1` for `lingerIterations`, but it still _has_ a penalty.
* Build a separate Java loop that does a much shorter sleep (5-10ms) and still uses `WNOHANG` with `waitpid`
  * This further shortens the penalty, at the expense of a bit more code
* Run a "tight loop" of `waitpid` calls with `WNOHANG` with no delay between them
  * This pays a lot of Java -> native -> Java overhead, but doesn't introduce artificial delays

The penalties of the different approaches are pretty obvious in the execution duration for the new `RunTest`:
* As implemented here on this pull request, on my machine the test (as in, all test cases) completes in ~800 _milliseconds_ (always less than 1 second)
* With `0` for iterations, it completes in 32 _seconds_
* With the new `protected` method, it completes in 18 _seconds_
* With the Java loop:
  * With a 10ms `Thread.sleep`, the test completes in 1.5 _seconds_
  * With a 50ms `Thread.sleep`, the test completes in 2.8 _seconds_
  * Ultimately, the penalty here will basically be whatever sleep we apply
* With the "tight loop", it completes in ~850ms
  * Repeated runs of this and the blocking `waitpid` on the pull request tend to show ~20-30ms is all that separates them; in practice, the penalty with the tight loop is pretty negligible

I picked this approach because it was the fastest and the simplest. However, it does come with the drawback that, if the process "never ends" (perhaps some sort of deadlock or other hang, which prevents it from ever terminating), this approach is basically "lost" forever waiting for it.

If we think that's a significant risk, I'd probably just go with the tight loop and pair that with some sort of either iteration count or maximum wait time where the loop "gives up". (One option for the "give up" path might be to belatedly register the process in the deadpool of one of the background processor threads, but the coordination for that is a bit complicated since the background threads are managed by the process factories rather than the processor types; e.g., `LinProcessFactory` manages the pool rather than `ProcessEpoll`)

We're happy to make any changes to use whatever approach you like best. I'm just trying to offer some details on the issue here and some insight into the reasoning for the change as included on the review.

Reply:
Re: "if the process "never ends" (perhaps some sort of deadlock or other hang, which prevents it from ever terminating)", it's worth bearing in mind that the process would have to hit that point after closing both `stdout` and `stderr`, to trigger the soft exit detection to add the process to the `deadPool`. If either of those is still open, then regardless of how we handle the `deadPool`, it won't matter because we won't get that far. This thread will still be in the `process()` loop in `run()`, waiting for the soft exit.